ios/BT/Extensions/Exposure Notifications/Scoring.swift

```diff
@@ -0,0 +1,55 @@
+import Foundation
+import ExposureNotification
+
+extension ExposureManager {
+
+  private static let matchedKeyCountMax = 3
+
+  static func score(summary: ENExposureDetectionSummary,
+                    with configuration: ExposureConfiguration) -> Bool {
+    if summary.matchedKeyCount <= matchedKeyCountMax {
+      // If the weighted average duration per matched key exceeds the
+      // threshold, we know that there was at least one day where the exposure
+      // exceeded the threshold.
+      let matchedKeyCount = Int(summary.matchedKeyCount)
+      if (matchedKeyCount > 0 &&
+        (weightedDuration(summary: summary,
+                          config: configuration) / matchedKeyCount) >=
+        configuration.triggerThresholdWeightedDuration * 60) {
+        return true
+      } else {
+        return false
+      }
+    } else {
+      // If the # matched keys is 4 or more, average duration is no longer
+      // useful because we have reached the cap of the attenuation buckets.
+      // (This assumes a trigger threshold of 15 minutes.)
+      if weightedDuration(summary: summary, config: configuration) >= configuration.maxWeightedDuration {
+        return true
+      } else {
+        return false
+      }
+    }
+  }
+
+  // Note: This is measured in seconds, while the config value is set in minutes and needs to be multiplied by 60.
+
+  static func weightedDuration(summary : ENExposureDetectionSummary,
+                               config: ExposureConfiguration) -> Int {
+    let weightedDuration =
+      config.attenuationBucketWeights[0] * summary.attenuationDurations[0].floatValue +
+      config.attenuationBucketWeights[1] * summary.attenuationDurations[1].floatValue +
+      config.attenuationBucketWeights[2] * summary.attenuationDurations[2].floatValue
+    return Int(weightedDuration)
+  }
+}
+
+extension ExposureConfiguration {
+
+  // Each bucket is capped at 30 minutes, this method calculates what the
+  // weighted duration ends up being capped at.
+
+  var maxWeightedDuration: Int {
+    return Int((attenuationBucketWeights[0] + attenuationBucketWeights[1] + attenuationBucketWeights[3]) * 30 * 60)
```

Comment:
Would be good to add a few simple unit tests to this class, that would've caught the [3] here.

Reply:
great catch, Lena. I was wondering what tests are needed for this scoring algorithm, like, edge cases and stuff, I am not really aware of what it does so any idea its appreciated

Reply:
Assuming a config with a threshold of 15 minutes, bucket weights 1, 0.5, 0, I would do exposures where:
- matched keys:1, attenuation bucket durations: 900, 0, 0: true. (weighted duration = 15 min because of low bucket)
- matched keys:1, attenuation bucket durations: 800, 0, 0: false. (weighted duration < 15 min)
- matched keys:1, attenuation bucket durations: 0, 900, 0: false. (weighted duration = 7.5 min)
- matched keys:1, attenuation bucket durations: 600, 600, 0: true. (weighted duration = 15 min)
- matched keys:1, attenuation bucket durations: 0, 0, 1800: false. (we don't use this bucket)

-matched keys: 2,  attenuation bucket durations: 900, 0, 0: false. (avg weighted duration = 7.5 min)
-matched keys: 2,  attenuation bucket durations: 900, 1800, 0: true. (avg weighted duration = 15min)

-matched keys: 3,  attenuation bucket durations: 900, 1800, 0: false. (avg weighted duration = 10min)
-matched keys: 3,  attenuation bucket durations: 1800, 1800, 0: true. (avg weighted duration = 15min)

-matched keys: 4, attenuation bucket durations: 900, 1800, 0: false. (max weighted duration not reached)
-matched keys: 4, attenuation bucket durations: 1800, 1800, 0: true. (max weighted duration reached)

Reply:
this is great, thanks!